internal/database/database.go

```diff
@@ -0,0 +1,150 @@
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Could not connect to database: %w", err)
+	}
+	return &MySqlDatabase{driver}, nil
+}
+
+func (db *MySqlDatabase) CreatePreAuthSession(csrfToken string, initTime time.Time) (*types.Session, error) {
+	// an empty account ID is functionally equivalent to NULL. Consider doing this
+	// instead of having account_id nullable?
+	return db.CreateSession("", csrfToken, initTime)
+}
+
+func (db *MySqlDatabase) CreateSession(accountId, csrfToken string, initTime time.Time) (*types.Session, error) {
+	sessionToken, err := crypto.GenerateToken()
+	if err != nil {
+		return nil, fmt.Errorf("Cound not generate token: %w", err)
+	}
+
+	// hooray more hardcoding!
+	const expireIdleTime = 30 * time.Minute
+	const expireAbsTime = 8 * time.Hour
+
+	stmt := "INSERT INTO Sessions(account_id, session_token, csrf_token, expire_idle, expire_abs) VALUES (?, ?, ?, ?, ?)"
+
+	// TODO: consider explicitly handling case where sessionToken happens to be a duplicate?
+	_, err = db.driver.Exec(stmt, accountId, sessionToken, csrfToken, initTime.Add(expireIdleTime), initTime.Add(expireAbsTime))
+	if err != nil {
+		return nil, fmt.Errorf("Error inserting new session: %w", err)
+	}
+
+	return &types.Session{accountId, sessionToken, csrfToken, initTime, initTime}, nil
+}
+
+func (db *MySqlDatabase) FetchSession(sessionToken string) (*types.Session, error) {
+	var accountIdRaw sql.NullString
+	var accountId, csrfToken string
+	var expireIdle, expireAbs time.Time
+
+	stmt := "SELECT account_id, csrf_token, expire_idle, expire_abs FROM Sessions WHERE session_token = ?"
+
+	err := db.driver.QueryRow(
+		stmt, sessionToken,
+	).Scan(&accountIdRaw, &csrfToken, &expireIdle, &expireAbs)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, fmt.Errorf("Session not found")
+		}
+
+		return nil, fmt.Errorf("SQL Error: %w", err)
+	}
+
+	if accountIdRaw.Valid {
+		accountId = accountIdRaw.String
+	}
+
+	return &types.Session{accountId, sessionToken, csrfToken, expireIdle, expireAbs}, nil
+}
+
+func (db *MySqlDatabase) DeleteSession(sessionToken string) error {
+	stmt := "DELETE FROM Sessions WHERE session_token = ?"
+	_, err := db.driver.Exec(stmt, sessionToken)
+	if err != nil {
+		return fmt.Errorf("Could not delete session: %w", err)
+	}
+	return nil
+}
+
+func (db *MySqlDatabase) FetchAccount(email string) (*types.Account, error) {
+	var accountId, passwordHash, salt string
+	stmt := "SELECT account_id, password_hash, salt FROM Accounts WHERE email = ?"
+	err := db.driver.QueryRow(
+		stmt, email,
+	).Scan(&accountId, &passwordHash, &salt)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, fmt.Errorf("No matching account found")
+		}
+		return nil, fmt.Errorf("SQL error: %w", err)
+	}
+
+	return &types.Account{accountId, email, passwordHash, salt}, nil
+}
+
+func (db *MySqlDatabase) DeleteExpiredSessions(t time.Time) error {
+	stmt := "DELETE FROM Sessions WHERE expire_idle < ? OR expire_abs < ?"
```

Comment:
Wouldn't this always delete any session that's existed for longer than 30 minutes (current expire_idle duration) - even if it's still "active"? I couldn't see expire_idle field being updated anywhere.

Reply:
expire_idle would be updated when a user makes an authenticated request. For the level 1 task, however, the only authenticated request is logout.